src/compiler/factory.ts

```diff
@@ -122,4 +122,210 @@ namespace ts {
+        // on the innermost expression in a chain of nested type assertions.
+        while (operand.kind === SyntaxKind.TypeAssertionExpression || operand.kind === SyntaxKind.AsExpression) {
+            operand = (<AssertionExpression>operand).expression;
+        }
+
+        // If the resulting expression is already parenthesized, we do not need to do any further processing.
+        if (operand.kind === SyntaxKind.ParenthesizedExpression) {
+            return operand;
+        }
+
+        return needsParenthesesForBinary(operand, operator, side)
+            ? parenthesizeExpression(operand)
+            : operand;
+    }
+
+    function needsParenthesesForBinary(operand: Expression, operator: SyntaxKind, side: BinaryOperand) {
+        const operandPrecedence = getExpressionPrecedence(operand);
+        const operatorPrecedence = getOperatorPrecedence(SyntaxKind.BinaryExpression, operator);
+        switch (compareValues(operandPrecedence, operatorPrecedence)) {
+            case Comparison.LessThan:
+                return true;
+            case Comparison.EqualTo:
+                return isRightAssociativeOperandOnLeftHandSide(operand, side)
+                    || isModuloOperandOnRightHandSide(operand, operator, side);
+            case Comparison.GreaterThan:
+                return false;
+        }
+    }
+
+    function isRightAssociativeOperandOnLeftHandSide(operand: Expression, side: BinaryOperand) {
+        return side === BinaryOperand.Left
+            && getExpressionAssociativity(operand) === Associativity.Right;
+    }
+
+    function isModuloOperandOnRightHandSide(operand: Expression, operator: SyntaxKind, side: BinaryOperand) {
+        return side === BinaryOperand.Right
+            && operator !== SyntaxKind.PercentToken
+            && operand.kind === SyntaxKind.BinaryExpression
+            && (<BinaryExpression>operand).operatorToken.kind === SyntaxKind.PercentToken;
+    }
+
+    function parenthesizeForAccess(expr: Expression): LeftHandSideExpression {
+        // When diagnosing whether the expression needs parentheses, the decision should be based
+        // on the innermost expression in a chain of nested type assertions.
+        while (expr.kind === SyntaxKind.TypeAssertionExpression || expr.kind === SyntaxKind.AsExpression) {
+            expr = (<AssertionExpression>expr).expression;
+        }
+
+        // isLeftHandSideExpression is almost the correct criterion for when it is not necessary
+        // to parenthesize the expression before a dot. The known exceptions are:
+        //
+        //    NewExpression:
+        //       new C.x        -> not the same as (new C).x
+        //    NumberLiteral
+        //       1.x            -> not the same as (1).x
+        //
+        if (isLeftHandSideExpression(expr) &&
+            expr.kind !== SyntaxKind.NewExpression &&
+            expr.kind !== SyntaxKind.NumericLiteral) {
+
+            return <LeftHandSideExpression>expr;

```

Comment:
Do you need the cast? I'd think `isLeftHandSideExpression` would be a guard on `expr`